File: java/source/src/main/java/com/yitter/idgen/DefaultIdGenerator.java
/*
 * 版权属于：yitter(yitter@126.com)
 * 开源地址：https://gitee.com/yitter/idgenerator
 */
package com.yitter.idgen;

import com.yitter.contract.ISnowWorker;
import com.yitter.contract.IdGeneratorException;
import com.yitter.contract.IdGeneratorOptions;
import com.yitter.contract.IIdGenerator;
import com.yitter.core.SnowWorkerM1;
import com.yitter.core.SnowWorkerM2;


public class DefaultIdGenerator implements IIdGenerator {

    private final ISnowWorker _SnowWorker;

    public DefaultIdGenerator(IdGeneratorOptions options) throws IdGeneratorException {
        if (options == null)
        {
            throw new IdGeneratorException("options error.");
        }

        if (options.StartTime > System.currentTimeMillis())
        {
            throw new IdGeneratorException("StartTime error.");
        }

        if (options.SeqBitLength + options.WorkerIdBitLength > 22)
        {
            throw new IdGeneratorException("error：WorkerIdBitLength + SeqBitLength <= 22");
        }

        double maxWorkerIdNumber = Math.pow(2, options.WorkerIdBitLength) - 1;
        if (options.WorkerId < 1 || options.WorkerId > maxWorkerIdNumber)
        {
            throw new IdGeneratorException("WorkerId error. (range:[1, " + maxWorkerIdNumber + "]");
        }

        if (options.SeqBitLength < 2 || options.SeqBitLength > 21)
        {
            throw new IdGeneratorException("SeqBitLength error. (range:[2, 21])");
        }

        double maxSeqNumber = Math.pow(2, options.SeqBitLength) - 1;
        if (options.MaxSeqNumber < 0 || options.MaxSeqNumber > maxSeqNumber)
        {
            throw new IdGeneratorException("MaxSeqNumber error. (range:[1, " + maxSeqNumber + "]");
        }

        double maxValue = maxSeqNumber - 2;
        if (options.MinSeqNumber < 5 || options.MinSeqNumber > maxValue)
        {
            throw new IdGeneratorException("MinSeqNumber error. (range:[5, " + maxValue + "]");
        }

        switch (options.Method)
        {
            case 1:
                _SnowWorker = new SnowWorkerM1(options);
                break;
            case 2:
                _SnowWorker = new SnowWorkerM2(options);
                break;
            default:
                _SnowWorker = new SnowWorkerM1(options);
                break;
        }

        if (options.Method == 1)
        {
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public long newLong()  {
        return _SnowWorker.nextId();
    }
}

File: Go/source/yitidgen/idgen/DefaultIdGenerator.go
/*
 * 版权属于：yitter(yitter@126.com)
 * 代码编辑：guoyahao
 * 代码修订：yitter
 * 开源地址：https://gitee.com/yitter/idgenerator
 */
package idgen

import (
	"../contract"
	"../core"
	"math"
	"time"
)

type DefaultIdGenerator struct {
	Options              *contract.IdGeneratorOptions
	SnowWorker           contract.ISnowWorker
	IdGeneratorException contract.IdGeneratorException
}

func NewDefaultIdGenerator(Options *contract.IdGeneratorOptions) *DefaultIdGenerator {
	if Options == nil {
		panic("dig.Options error.")
	}

	var minTime = time.Now().AddDate(-50, 0, 0).UnixNano() / 1e6
	if minTime == 0 || Options.BaseTime < minTime || Options.BaseTime > time.Now().UnixNano()/1e6 {
		panic("BaseTime error.")
	}

	if Options.SeqBitLength+Options.WorkerIdBitLength > 22 {
		panic("error：WorkerIdBitLength + SeqBitLength <= 22")
	}

	maxWorkerIdNumber := uint16(math.Pow(float64(2), float64(Options.WorkerIdBitLength))) - 1
	if Options.WorkerId > maxWorkerIdNumber {
		panic("WorkerId error. (range:[1, "+ string(maxWorkerIdNumber)+ "]")
	}

	if Options.SeqBitLength < 2 || Options.SeqBitLength > 21 {
		panic("SeqBitLength error. (range:[2, 21])")
	}

	maxSeqNumber := uint32(math.Pow(2, float64(Options.SeqBitLength))) - 1
	if Options.MaxSeqNumber > maxSeqNumber {
		panic("MaxSeqNumber error. (range:[1, "+ string(maxSeqNumber)+ "]")
	}

	if Options.MinSeqNumber > maxSeqNumber {
		panic("MinSeqNumber error. (range:[1, "+ string(maxSeqNumber)+ "]")
	}

	var snowWorker contract.ISnowWorker

	switch Options.Method {
	case 1:
		snowWorker = core.NewSnowWorkerM1(Options)
	case 2:
		snowWorker = core.NewSnowWorkerM2(Options)
	default:
		snowWorker = core.NewSnowWorkerM1(Options)
	}

	if Options.Method == 1 {
		time.Sleep(time.Duration(500) * time.Microsecond)
	}
	return &DefaultIdGenerator{
		Options:    Options,
		SnowWorker: snowWorker,
	}
}

func (dig DefaultIdGenerator) NewLong() uint64 {
	return dig.SnowWorker.NextId()
}
